File: publishing-worker/src/publishers/wechat.ts
import type {
  PlatformCredentials,
  PlatformPublisher,
  PublicationInput,
  PublicationResult,
  PublicationStatusInput,
  PublicationStatusResult,
} from "./types.js";

type WechatResponse = Record<string, unknown> & { errcode?: number; errmsg?: string };

async function jsonRequest(url: string, init?: RequestInit): Promise<WechatResponse> {
  const response = await fetch(url, init);
  if (!response.ok) throw new Error("platform_unavailable");
  const data = (await response.json()) as WechatResponse;
  if (typeof data.errcode === "number" && data.errcode !== 0) {
    const authCodes = new Set([40001, 40014, 42001]);
    throw new Error(authCodes.has(data.errcode) ? "authorization_required" : "platform_rejected_request");
  }
  return data;
}

async function accessToken(credentials: PlatformCredentials) {
  if (credentials.access_token) return credentials.access_token;
  if (!credentials.app_id || !credentials.app_secret) throw new Error("authorization_required");
  const data = await jsonRequest(
    `https://api.weixin.qq.com/cgi-bin/token?grant_type=client_credential&appid=${encodeURIComponent(credentials.app_id)}&secret=${encodeURIComponent(credentials.app_secret)}`,
  );
  const token = typeof data.access_token === "string" ? data.access_token : "";
  if (!token) throw new Error("authorization_required");
  return token;
}

async function fetchImage(url: string) {
  const response = await fetch(url, { redirect: "follow" });
  if (!response.ok) throw new Error("media_download_failed");
  const bytes = await response.arrayBuffer();
  if (!bytes.byteLength || bytes.byteLength > 10 * 1024 * 1024) throw new Error("media_invalid");
  return { bytes, type: response.headers.get("content-type") || "image/jpeg" };
}

async function uploadMaterial(token: string, url: string, type: "thumb" | "image") {
  const image = await fetchImage(url);
  const form = new FormData();
  form.append("media", new Blob([image.bytes], { type: image.type }), type === "thumb" ? "cover.jpg" : "image.jpg");
  const endpoint = type === "thumb"
    ? `https://api.weixin.qq.com/cgi-bin/material/add_material?access_token=${encodeURIComponent(token)}&type=thumb`
    : `https://api.weixin.qq.com/cgi-bin/media/uploadimg?access_token=${encodeURIComponent(token)}`;
  const response = await fetch(endpoint, { method: "POST", body: form });
  if (!response.ok) throw new Error("platform_unavailable");
  const data = (await response.json()) as WechatResponse;
  if (typeof data.errcode === "number" && data.errcode !== 0) throw new Error("media_invalid");
  return data;
}

async function getPublishStatus(token: string, publishId: string) {
  return jsonRequest(
    `https://api.weixin.qq.com/cgi-bin/freepublish/get?access_token=${encodeURIComponent(token)}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ publish_id: publishId }),
    },
  );
}

function publicArticleUrl(status: WechatResponse) {
  const detail = status.article_detail as { item?: Array<{ article_url?: string }> } | undefined;
  return detail?.item?.find((item) => typeof item.article_url === "string")?.article_url;
}

export class WechatPublisher implements PlatformPublisher {
  readonly platformKey = "wechat";
  // 正式 API 流程已接入，但必须用显问测试公众号完成真实发布验收后才能标记为已验证。
  readonly verifiedCapabilities = [] as const;

  async checkAuth(credentials: PlatformCredentials) {
    try {
      await accessToken(credentials);
      return { ok: true };
    } catch {
      return { ok: false };
    }
  }

  async checkStatus(input: PublicationStatusInput): Promise<PublicationStatusResult> {
    if (!input.externalPostId) {
      return { platformKey: this.platformKey, status: "unknown" };
    }
    try {
      const token = await accessToken(input.credentials);
      const status = await getPublishStatus(token, input.externalPostId);
      const publishStatus = typeof status.publish_status === "number" ? status.publish_status : -1;
      if (publishStatus === 0) {
        const publicUrl = publicArticleUrl(status);
        return publicUrl
          ? { platformKey: this.platformKey, status: "published", publicUrl }
          : { platformKey: this.platformKey, status: "submitted" };
      }
      if ([2, 3, 5, 6].includes(publishStatus)) {
        return { platformKey: this.platformKey, status: "failed", safeErrorCode: "content_rejected" };
      }
      return { platformKey: this.platformKey, status: "submitted" };
    } catch (error) {
      const code = error instanceof Error ? error.message : "platform_unavailable";
      if (code === "authorization_required") {
        return { platformKey: this.platformKey, status: "auth_required", safeErrorCode: code };
      }
      return { platformKey: this.platformKey, status: "unknown", safeErrorCode: "platform_unavailable" };
    }
  }

  async publish(input: PublicationInput): Promise<PublicationResult> {
    try {
      const token = await accessToken(input.credentials);
      const cover = input.assets.find((item) => item.role === "cover") || input.assets[0];
      if (!cover) {
        return { success: false, platformKey: this.platformKey, status: "failed", safeErrorCode: "media_invalid" };
      }
      const coverResult = await uploadMaterial(token, cover.url, "thumb");
      const thumbMediaId = typeof coverResult.media_id === "string" ? coverResult.media_id : "";
      if (!thumbMediaId) return { success: false, platformKey: this.platformKey, status: "failed", safeErrorCode: "media_invalid" };

      let content = input.contentHtml;
      for (const asset of input.assets.filter((item) => item.role !== "cover").slice(0, 8)) {
        try {
          const uploaded = await uploadMaterial(token, asset.url, "image");
          const url = typeof uploaded.url === "string" ? uploaded.url : "";
          if (url) content += `<p><img src="${url}" alt="${(asset.alt || "").replace(/[<>\"]/g, "")}" /></p>`;
        } catch {
          // 单张正文插图失败不应破坏主文章；封面失败才阻断发布。
        }
      }

      const draft = await jsonRequest(
        `https://api.weixin.qq.com/cgi-bin/draft/add?access_token=${encodeURIComponent(token)}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            articles: [
              {
                title: input.title.slice(0, 64),
                author: "",
                digest: (input.summary || input.contentText).slice(0, 120),
                content,
                content_source_url: "",
                thumb_media_id: thumbMediaId,
                need_open_comment: 0,
                only_fans_can_comment: 0,
              },
            ],
          }),
        },
      );
      const mediaId = typeof draft.media_id === "string" ? draft.media_id : "";
      if (!mediaId) return { success: false, platformKey: this.platformKey, status: "failed", safeErrorCode: "platform_unavailable" };
      if (input.publishMode === "draft") {
        return { success: true, platformKey: this.platformKey, status: "drafted", externalPostId: mediaId };
      }

      const submitted = await jsonRequest(
        `https://api.weixin.qq.com/cgi-bin/freepublish/submit?access_token=${encodeURIComponent(token)}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ media_id: mediaId }),
        },
      );
      const publishId = typeof submitted.publish_id === "string" ? submitted.publish_id : "";
      if (!publishId) return { success: false, platformKey: this.platformKey, status: "failed", safeErrorCode: "platform_unavailable" };

      for (let attempt = 0; attempt < 4; attempt += 1) {
        await new Promise((resolve) => setTimeout(resolve, 1800));
        const status = await getPublishStatus(token, publishId);
        const publishStatus = typeof status.publish_status === "number" ? status.publish_status : -1;
        if (publishStatus === 0) {
          const publicUrl = publicArticleUrl(status);
          if (publicUrl) {
            return { success: true, platformKey: this.platformKey, status: "published", externalPostId: publishId, publicUrl };
          }
          return { success: true, platformKey: this.platformKey, status: "submitted", externalPostId: publishId };
        }
        if ([2, 3, 5, 6].includes(publishStatus)) {
          return { success: false, platformKey: this.platformKey, status: "failed", externalPostId: publishId, safeErrorCode: "content_rejected" };
        }
      }
      return { success: true, platformKey: this.platformKey, status: "submitted", externalPostId: publishId };
    } catch (error) {
      const code = error instanceof Error ? error.message : "platform_unavailable";
      if (code === "authorization_required") return { success: false, platformKey: this.platformKey, status: "auth_required", safeErrorCode: code };
      return { success: false, platformKey: this.platformKey, status: "failed", safeErrorCode: code === "media_invalid" ? "media_invalid" : "platform_unavailable" };
    }
  }
}

File: backend/apps/publications/wechat.py
from __future__ import annotations

import html
import time
from dataclasses import dataclass

import httpx

from apps.documents.storage import storage_provider

from .models import PublicationTarget


class WeChatCredentialError(RuntimeError):
    pass


class WeChatPublishError(RuntimeError):
    pass


@dataclass(frozen=True)
class WeChatPublishResult:
    status: str
    external_post_id: str = ""
    public_url: str = ""


def _json(response: httpx.Response) -> dict:
    try:
        value = response.json()
    except Exception as exc:
        raise WeChatPublishError("invalid platform response") from exc
    if not isinstance(value, dict):
        raise WeChatPublishError("invalid platform response")
    return value


def _access_token(credentials: dict) -> str:
    app_id = str(credentials.get("app_id", "")).strip()
    app_secret = str(credentials.get("app_secret", "")).strip()
    if not app_id or not app_secret:
        raise WeChatCredentialError("missing official credentials")
    with httpx.Client(timeout=15.0) as client:
        response = client.get(
            "https://api.weixin.qq.com/cgi-bin/token",
            params={"grant_type": "client_credential", "appid": app_id, "secret": app_secret},
        )
    data = _json(response)
    token = str(data.get("access_token", "")).strip()
    if not token:
        raise WeChatCredentialError(str(data.get("errcode", "credential rejected")))
    return token


def validate_wechat_credentials(credentials: dict) -> dict:
    _access_token(credentials)
    return {"validated": True}


def _cover_bytes(target: PublicationTarget):
    visual = (
        target.job.visuals.select_related("image")
        .filter(target__isnull=True, role="cover")
        .order_by("ordinal", "created_at")
        .first()
    )
    if visual is None:
        raise WeChatPublishError("cover required")
    image = visual.image
    with storage_provider().open_object(image.object_key) as source:
        data = source.read()
    if not data:
        raise WeChatPublishError("cover unavailable")
    extension = {"image/png": "png", "image/jpeg": "jpg", "image/webp": "webp"}.get(
        image.mime_type, "jpg"
    )
    return data, image.mime_type, f"cover.{extension}"


def _upload_cover(token: str, target: PublicationTarget) -> str:
    data, mime, filename = _cover_bytes(target)
    with httpx.Client(timeout=30.0) as client:
        response = client.post(
            "https://api.weixin.qq.com/cgi-bin/material/add_material",
            params={"access_token": token, "type": "thumb"},
            files={"media": (filename, data, mime)},
        )
    payload = _json(response)
    media_id = str(payload.get("media_id", "")).strip()
    if not media_id:
        raise WeChatPublishError(f"cover upload failed:{payload.get('errcode', '')}")
    return media_id


def _article_html(content: str) -> str:
    paragraphs = [value.strip() for value in content.replace("\r", "").split("\n") if value.strip()]
    return "".join(f"<p>{html.escape(value)}</p>" for value in paragraphs)


def _add_draft(token: str, target: PublicationTarget, thumb_media_id: str) -> str:
    payload = target.payload_snapshot or {}
    title = str(payload.get("title", "")).strip()[:64]
    content = str(payload.get("content", "")).strip()
    if not title or not content:
        raise WeChatPublishError("content unavailable")
    request_body = {
        "articles": [
            {
                "title": title,
                "author": "",
                "digest": content.replace("\n", " ")[:120],
                "content": _article_html(content),
                "content_source_url": "",
                "thumb_media_id": thumb_media_id,
                "need_open_comment": 0,
                "only_fans_can_comment": 0,
            }
        ]
    }
    with httpx.Client(timeout=20.0) as client:
        response = client.post(
            "https://api.weixin.qq.com/cgi-bin/draft/add",
            params={"access_token": token},
            json=request_body,
        )
    data = _json(response)
    media_id = str(data.get("media_id", "")).strip()
    if not media_id:
        raise WeChatPublishError(f"draft failed:{data.get('errcode', '')}")
    return media_id


def _submit_publish(token: str, media_id: str) -> str:
    with httpx.Client(timeout=20.0) as client:
        response = client.post(
            "https://api.weixin.qq.com/cgi-bin/freepublish/submit",
            params={"access_token": token},
            json={"media_id": media_id},
        )
    data = _json(response)
    publish_id = str(data.get("publish_id", "")).strip()
    if not publish_id:
        raise WeChatPublishError(f"publish submit failed:{data.get('errcode', '')}")
    return publish_id


def check_wechat_publish(credentials: dict, publish_id: str) -> WeChatPublishResult:
    token = _access_token(credentials)
    with httpx.Client(timeout=20.0) as client:
        response = client.post(
            "https://api.weixin.qq.com/cgi-bin/freepublish/get",
            params={"access_token": token},
            json={"publish_id": publish_id},
        )
    data = _json(response)
    status = data.get("publish_status")
    if status == 1:
        return WeChatPublishResult(status="pending", external_post_id=publish_id)
    if status != 0:
        raise WeChatPublishError(f"publish failed:{status}:{data.get('fail_idx', '')}")
    detail = data.get("article_detail") or {}
    items = detail.get("item") or []
    public_url = ""
    if isinstance(items, list) and items and isinstance(items[0], dict):
        public_url = str(items[0].get("article_url") or items[0].get("url") or "")
    return WeChatPublishResult(
        status="published", external_post_id=publish_id, public_url=public_url
    )


def publish_wechat_target(target: PublicationTarget, credentials: dict) -> WeChatPublishResult:
    if target.external_post_id:
        return check_wechat_publish(credentials, target.external_post_id)
    token = _access_token(credentials)
    thumb_media_id = _upload_cover(token, target)
    draft_id = _add_draft(token, target, thumb_media_id)
    publish_id = _submit_publish(token, draft_id)
    for _ in range(4):
        time.sleep(2)
        result = check_wechat_publish(credentials, publish_id)
        if result.status == "published":
            return result
    return WeChatPublishResult(status="pending", external_post_id=publish_id)
